extensions/amp-pan-zoom/1.0/component-ts.tsx

```diff
@@ -0,0 +1,267 @@
+import {
+  GestureHandlers,
+  UserGestureConfig,
+  UserHandlers,
+  useGesture,
+} from '@use-gesture/react';
+import type {ComponentChildren, Ref} from 'preact';
+
+import {scale as cssScale, px, translate} from '#core/dom/style';
+
+import * as Preact from '#preact';
+import {useImperativeHandle, useRef} from '#preact';
+import {forwardRef} from '#preact/compat';
+import {ContainWrapper} from '#preact/component';
+import {useResizeObserver} from '#preact/hooks/useResizeObserver';
+
+import {useStyles} from './component.jss';
+import {usePanZoomState} from './hooks/usePanZoomState';
+
+const DOUBLE_TAP_TIME = 500; // Maximum time between tap starts
+
+const gestureConfig: Omit<UserGestureConfig, 'target'> = {
+  drag: {
+    filterTaps: true,
+    tapsThreshold: 10,
+  },
+};
+
+export type BentoPanZoomProps = {
+  children?: ComponentChildren;
+  controls?: boolean;
+  initialScale?: number | string;
+  initialX?: number | string;
+  initialY?: number | string;
+  maxScale?: number | string;
+};
+
+export type BentoPanZoomApi = {
+  transform(scale: number, x: number, y: number): void;
+};
+
+function getElementPosition(
+  clientX: number,
+  clientY: number,
+  element: HTMLElement
+) {
+  const elBounds = element./* REVIEW */ getBoundingClientRect();
+  return {
+    anchorX: clientX - elBounds.x,
+    anchorY: clientY - elBounds.y,
+  };
+}
+
+function classNames(...args: Array<string | false | null | 0>) {
+  return args.filter(Boolean).join(' ');
+}
+
+/**
+ * @return {PreactDef.Renderable}
+ */
+export function BentoPanZoomWithRef(
+  {
+    children,
+    controls = true,
+    initialScale,
+    initialX,
+    initialY,
+    maxScale,
+    ...rest
+  }: BentoPanZoomProps,
+  ref: Ref<BentoPanZoomApi>
+) {
+  const styles = useStyles();
+
+  const [state, actions] = usePanZoomState({
+    initialX,
+    initialY,
+    initialScale,
+    maxScale,
+  });
+  const isPannable = state.scale !== 1;
+  const canZoom = state.scale < state.maxScale;
+
+  const contentRef = useRef<HTMLDivElement>(null);
+  const containerRef = useRef<HTMLDivElement>(null);
+
+  const updateSizes = () => {
+    const contentBox = contentRef.current!./* REVIEW */ getBoundingClientRect();
+    const containerBox =
+      containerRef.current!./* REVIEW */ getBoundingClientRect();
```

Comment:
Isn't this available from the ResOb?

Reply:
The ResOb only reports size metrics (`contentBoxSize: [ width, height]`) .  There is a deprecated property called `contentRect` which is probably the same rect that `getBoundingClientRect` gives us, but in this scenario we always need to know the x/y of  both elements.
Fortunately, since the ResizeObserver only calls this callback after a layout, these calls should not perform a reflow.